mesh_segmenter/src/mesh_segmenter.cpp

```diff
@@ -62,29 +64,65 @@ std::vector<vtkSmartPointer<vtkPolyData> > MeshSegmenter::getMeshSegments()
 
 void MeshSegmenter::segmentMesh()
 {
+  // Cells that are in included_indices from the segmentation
   vtkSmartPointer<vtkIdList> used_cells = vtkSmartPointer<vtkIdList>::New();
+  // Cells that were rejected because the cluster they were in did not satisfy the parameters defined
+  vtkSmartPointer<vtkIdList> rejected_cells = vtkSmartPointer<vtkIdList>::New();
   int size = input_mesh_->GetCellData()->GetNumberOfTuples();
 
   used_cells->Allocate(size);
+  rejected_cells->Allocate(size);
 
   included_indices_.clear();
 
-  for(int i = 0; i < size; ++i)
+  // Loop over all of the cells
+  for (int i = 0; i < size; ++i)
   {
     // if current index is not used, perform segmentation with it and insert the results into the list
-    if(used_cells->IsId(i) == -1)
+    if (used_cells->IsId(i) == -1)
     {
       vtkSmartPointer<vtkIdList> linked_cells = vtkSmartPointer<vtkIdList>::New();
       linked_cells = segmentMesh(i);
 
       // save indices found
-      included_indices_.push_back(linked_cells);
-      for(int j = 0; j < linked_cells->GetNumberOfIds(); ++j)
+      if (linked_cells->GetNumberOfIds() > min_cluster_size_)
+      {
+        included_indices_.push_back(linked_cells);
+
+        for (int j = 0; j < linked_cells->GetNumberOfIds(); ++j)
+        {
+          used_cells->InsertNextId(linked_cells->GetId(j));
+        }
+      }
+      else
       {
-        used_cells->InsertNextId(linked_cells->GetId(j));
+        for (int j = 0; j < linked_cells->GetNumberOfIds(); ++j)
+        {
+          rejected_cells->InsertNextId(linked_cells->GetId(j));
+        }
       }
     }
   }
+
+  // Anything still not used in a segment is an "edge" - There has to be an easier way...
+  vtkSmartPointer<vtkIdList> edge_cells = vtkSmartPointer<vtkIdList>::New();
+  for (int i = 0; i < size; i++)
+  {
+    // If ID i is not in the used_cells list (and thus isn't in included_indices) and is a reject (not sure what the
+    // alternative would be)
+    if ((used_cells->IsId(i) == -1) && (rejected_cells->IsId(i) != -1))
+    {
+      // Then that cell is an "edge". Note - there is still something odd here. If you have trouble with it, try looking
+      // at the IDs you are getting inserting for each i
+      edge_cells->InsertNextId(rejected_cells->GetId(i));
+    }
+  }
+  included_indices_.push_back(edge_cells);
+
+  std::cout << "Found " << included_indices_.size() << " segments" << '\n';
```

Comment:
OK, my understanding is that you use cout in the computational/non-ros code and ROS_ERROR etc for ROS code

Reply:
That's mine as well. In this case this is non-ros code. ROS_ERROR will not compile. I could also just take them out if you prefer.